### DISCUSSION TITLE
Works in dev but failing in production

### DISCUSSION THREAD

**Discussion by ilanmazuz** [RehanSaeed/Schema.NET#185] [Q&A]
Hello and thank you so much for this project, it is a great help.

In our asp.net ajax project, we've installed the Nuget package in the dev machine and were able to generate the json of an faqPage. 
However, when we added the dll and its dependencies to the production server, the first line that refers the Schema.net classes in code cause an error which even a try/catch block does not log and we have no idea where to start.

we have registered all the dependencies in the GAC and added the assemblies to the web.config :

				<add assembly="netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51"/>
				<add assembly="System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />

				<add assembly="Newtonsoft.Json, Version=9.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed" />
				<add assembly="Schema.NET, Version=7.1.0.0, Culture=neutral, PublicKeyToken=fc5550082a9c642c"/>
				<add assembly="System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51"/>
				<add assembly="System.Memory, Version=4.0.1.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51"/>
				<add assembly="System.Numerics.Vectors, Version=4.1.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
				<add assembly="System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"/>

p.s. the first assembly "netstandard" was not referred to in the nuget or anywhere, but we were forced to add it even in the dev machine otherwise nothing could load and the most basic .net objects stopped working and throws error like "unknown class".

the difference between the dev and production is the fact that the dll's are in the Bin folder and not as a reference in the web.config (except for the first 2 lines which are the same). i thought maybe we are referring a wrong version of some of the dlls (the version number of their build differ from the version number they are actually registered with in the GAC, we checked the GAC to know which version to use and which PublicKeyToken to have for each of them),
However, maybe we refer a wrong version in some of them, for example the System.Numerics.Vectors have 2 versions installed, 4.0.0.0 and 4.1.3.0 - which we referred to, but maybe we need to use the older version?
see here:
![image](https://user-images.githubusercontent.com/8776013/87292458-8120a800-c509-11ea-9751-75fc2c9d7fab.png)
the problem a i mentioned is that the dll files themselves shows completely different version... see here:
![image](https://user-images.githubusercontent.com/8776013/87292768-f8eed280-c509-11ea-8036-3992061dae52.png)

So like i said, when the line:
var faqPage = new FAQPage();
is reached - the production site is redirecting us to the homepage because of some internal error which we don't get to see in the try/catch log at all.
When we comment the code lines that refer to Schema.net, the faq page is loading fine.

please advise.

Thanks a lot,
ilan.

**Comment by ilanmazuz** [comment]
Hello again,

i was able to confirm that the system.numeric.vectors version which i installed with the Schma.net last week (on July 9th 2020) is indeed 4.1.3.0 as you can see here:
![image](https://user-images.githubusercontent.com/8776013/87295007-3d2fa200-c50d-11ea-900c-daf88a496979.png)

all the other dll's does not have another version installed at all, so the issue is not a wrong version.

I hope you have an idea what is wrong.

Thanks again,
ilan.

**Comment by RehanSaeed** [comment]
Looks like you are using .NET Framework and are having binding redirect issues. This project targets `net461` and `net472`, so in theory everything should work. You probably need to figure out how to get the web.config right. Sorry I can't be of more help.

**Comment by ilanmazuz** [comment]
Thank you for the quick reply. 
on the server we have net4.8 installed, will it work?
![image](https://user-images.githubusercontent.com/8776013/87331180-425b1400-c542-11ea-8b85-2370c9c9bf97.png)

Anyway, in the web.config it seems we still point to version 4.0 (<compilation debug="false" optimizeCompilations="true" defaultLanguage="c#" targetFramework="4.0" >). So tomorrow during off hours i will change it to 4.8 and see if it solved the problem. i will post back afterwards.

Thanks a lot.

**Comment by ilanmazuz** [comment]
Hi,
In the past hour i tried to set the target framework to 4.8 or 4.7.2 but it did not help.
![image](https://user-images.githubusercontent.com/8776013/87395663-66f2d280-c5ba-11ea-92d4-c5802d9a6e8b.png)

i tried to set compilation debug="true" and i have <customErrors mode="RemoteOnly" in hope to see anything in the gaq page that will explain what is the error, but it always keep redirecting me to the domain homepage, which i have no idea why.

You suggested that is has something to to with Binding Redirect, but Binding Redirect has nothing to do with page redirection, only with "dll version redirecting", right? so it can't be this.

did anything in my web.config assemblies, which i posted in my first post, seems wrong to you? like i said - in the dev we are placing the dll's in the Bin folder and not in GAC - so maybe i am doing something wrong there?

Thanks,
ilan.

**Comment by RehanSaeed** [comment]
One thing that might help is if you try to create a minimum viable reproducable project. So maybe start with a blank project and keep adding things until you hit the problem.

**Comment by ilanmazuz** [comment]
Thanks for the advise.
I was able to pinpoint the problem to the NewtonSoft.Json version after all.
As i showed in the first post, the dll installed with Nuget is showing version 12.0.3.23909 but when registering it in GAC is listed as version 9.0.0.0 (i triple checked it) and that's the only way the Web.Config is able to recognize it, but your Schema.Net dll requires that the V12 will be used...
The only solution was to add a Bin folder to the production project and place the Newtonsoft dll there and remove the reference from the web.config to the GAC assembly.

**Comment by RehanSaeed** [comment]
Cool. Thanks for posting your answer.
